### DISCUSSION TITLE
xrdp reconnect behaviour (xrdp 0.9.x)

### DISCUSSION THREAD

**Discussion by JAMESADEV** [neutrinolabs/xrdp#2717] [Q&A]
Hi, 
I use xrdp 0.9.17 on a ubuntu 22.04 host. From time to time my xrdp session has some issues. For example audio or clipboard will stop working. A restart of the xrdp service solves the issue. However, whenever I restart xrdp, it creates a new Xorg display when my rdp client reconnects. My understanding is that for a while now xrdp has been able to allow rdp clients to reconnect to an existing session.  Is the new Xorg display my xrdp server creates a bug or is it expected behaviour after a "systemctl restart xrdp"?

If is not a bug, is there any advice on how to make sure my rdp client always connects to Xorg :10 and stop xrdp from creating Xorg :11 :12 etc?

Thank you.

**Comment by matt335672** [comment]
There are two known problems you're hitting here:-

1. When xrdp is restarted, the ability to connect to existing sessions is lost. At the moment that's expected behaviour. We're aware of it however, and it will be addressed in a future release.
2. When a client connects to a session which already has a client, the new connection is unable to connect to a component called chansrv which services (among other things) the audio and clipboard. A future version of xrdp will terminate the first connection when a second connection comes in, avoiding this problem. 

There's a workaround for 2, but it's fiddly. If you log in and are unable to connect to the clipboard and/or audio:-
- Use `ps -ef | grep xrdp` to locate all the xrdp processes. On Ubuntu these run as the `xrdp` user.
- One of the processes  (called the listener) will have a parent PID of 1. The others will have a parent pid of the listener process.
- Use `sudo kill <pidlist>` to kill all the xrdp processes except the listener. That will log out all user sessions.
- When you next log in, all should be well.

Here's a minimally engineered script which you can try which should do all the above for you. Copy-paste this to `kill_all_xrdp.sh` and try running it :-

```sh
#!/bin/sh

# Script to kill all xrdp processes except the listener

# Username xrdp runs as
XRDP_USER=xrdp

if ! getent passwd "$XRDP_USER" >/dev/null; then
    echo "** xrdp user '$XRDP_USER' does not exist on this system" >&2
    exit 1
fi

pidlist=

for pid in `pgrep -u $XRDP_USER xrdp\$`; do
    ppid=`ps -o ppid= -p $pid` ; # Get parent pid
    if [ -z "$ppid" ]; then
        echo "** Can't get Parent PID of $pid" >&2
    elif [ $ppid -ne 1 ]; then
        pidlist="$pidlist $pid"
    fi
done

if [ -z "$pidlist" ]; then
    echo "** No xrdp sessions were found" >&2
    false
else
    echo "Killing PIDs :$pidlist"
    sudo -u $XRDP_USER kill $pidlist
fi

exit $?
```

**Reply by JAMESADEV** [reply]
Hi Matt, thank you. I will try this and let you know. Is the a rough ETA on the fix for 1?

**Reply by matt335672** [reply]
Not at all I'm afraid. It'll be in the lifetime of the next major release. We've just redesigned the sesman architecture to support a restart, but there's a way to go yet.
